bpo-22281 ENH add introspection API for concurrent.futures Executor

tomMoral opened the pull request:
This PR is building from the patch coming from the orignal issue (https://bugs.python.org/issue22281) [GH-66477] to add an introspection API for the `concurrent.futures` executors. The following methods are added:

* `worker_count()` : Total number of workers currently in the pool
* `active_worker_count()` : Number of workers currently processing a work item
* `idle_worker_count()`: Number of workers not processing a work item
---
* `task_count()`: Total number of tasks currently being handled by the pool
* `active_task_count()`: Number of tasks currently being processed by workers
* `waiting_task_count()`: Number of submitted tasks not yet being processed by a worker
---
* `active_tasks()`: A set of _WorkItem objects currently being processed by a worker.
* `waiting_tasks()`: A list of `_WorkItem` objects currently waiting to be processed by a worker.

The discussion in the orignal issue proposed to use properties instead of methods and to add a couple functionalities. I am up for inputs on that. I think a nice addition would be some indication of the work load in the `Executor`, by timing the idle time and active time of each tasks and collecting it.

<!-- issue-number: bpo-22281 -->
https://bugs.python.org/issue22281
<!-- /issue-number -->

tomMoral committed:
ENH add introspection API for concurrent.futures Executor

pitrou commented:
As I said in the original issue, I think this should be a single `stats()` method that returns a dictionary. There are two reasons:
* With several methods, there's a race condition: if I call `waiting_task_count()` then `active_task_count()`, I'm not sure the two numbers are comparable since they reflect snapshots at different points in time; a single method can guarantee the snapshot is consistent (perhaps it needs to take a lock for that, that's an implementation issue)
* It makes the API simpler: we can add a lot of content to the stats dictionary without clobbering the API and the API docs :-)

Of course, the returned dictionary would be something like:
```
{'worker_count': 8,
 'active_worker_count': 3,
 'idle_worker_count': 5,
 # etc.
}
```
I'd also see something like a `"status`" field reflecting the state of the executor ("running", "shutting_down", "shutdown", "broken" perhaps).

pitrou commented:
By the way, you should probably base your work on PR #4241 because I'm soon gonna merge it :-)

tomMoral commented:
I just rebased the PR and I will implement the `stat` API asap.

pitrou commented:
Woops... It looks like squashing PR #4241 on push left this PR with conflicts. I'm sorry :-/

tomMoral committed:
ENH change introspect API to

tomMoral committed:
ENH add status in the introspect API

- Use a ExecutorFlags object for more consistent status
- Improve the shutdown of the ProcessPoolExecutor

tomMoral committed:
DOC update the doc to contain 'status'
